File: scripts/sync_all_supabase_orders_with_shalom.py
import urllib.request
import json
import ssl
import re

ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

SUPABASE_URL = "https://uwmdjsxwetjvsxsdngko.supabase.co"
ANON_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InV3bWRqc3h3ZXRqdnN4c2RuZ2tvIiwicm9sZSI6ImFub24iLCJpYXQiOjE3ODY2NDE5MTEsImV4cCI6MjEwMjIxNzkxMX0.KaqryIyoe4IDQGTJD_cswZkW-wfgnMcyV9tJoWxHMq8"
API_KEY = "sk_qm4rm5ivepety4ausqnubkfegp4yr2lnqu3p4q55oc3v4yzw3oma"
BASE_URL = "https://api.shalom-api-peru.com"

# 1. Obtener todas las órdenes de Shalom Pro (página 1 y 2)
shalom_orders = []
for p in [1, 2]:
    req_sh = urllib.request.Request(f"{BASE_URL}/v1/orders?per_page=100&page={p}", headers={
        "X-API-Key": API_KEY,
        "X-Shalom-Email": "milagrosjanetamis@gmail.com",
        "X-Shalom-Password": "986398Mi$",
        "User-Agent": "Mozilla/5.0"
    })
    try:
        with urllib.request.urlopen(req_sh, context=ctx, timeout=15) as resp:
            data = json.loads(resp.read().decode('utf-8'))
            items = data.get('data') or data.get('orders') or []
            shalom_orders.extend(items)
    except Exception as e:
        print(f"[-] Error pagina {p}: {e}")

print(f"[*] Total ordenes cargadas de Shalom Pro: {len(shalom_orders)}")

# 2. Obtener pedidos de Supabase
req_p = urllib.request.Request(f"{SUPABASE_URL}/rest/v1/pedidos?select=*&order=created_at.desc&limit=50", headers={
    "apikey": ANON_KEY,
    "Authorization": f"Bearer {ANON_KEY}",
    "User-Agent": "Mozilla/5.0"
})
with urllib.request.urlopen(req_p, context=ctx, timeout=10) as resp:
    pedidos = json.loads(resp.read().decode('utf-8'))

print(f"[*] Obtenidos {len(pedidos)} pedidos de Supabase")

synced_count = 0

for p in pedidos:
    destino = p.get('destino_detalle') or ''
    m = re.search(r'(?:DNI/CE|DNI|CE|Doc|Documento)[\s:]*(?:Recojo:?\s*)?([A-Za-z0-9]{6,12})', destino, re.IGNORECASE)
    dni = m.group(1).strip() if m else ''
    
    if not dni:
        continue
    
    # Buscar en órdenes de Shalom Pro
    matched = None
    for sh in shalom_orders:
        rec = sh.get('receiver') or sh.get('destinatario') or {}
        rec_doc = str(rec.get('document') or rec.get('documento') or '').strip()
        if rec_doc == dni or (len(dni) >= 8 and rec_doc.endswith(dni)):
            matched = sh
            break
            
    if matched:
        full_guia = f"{matched.get('serie') or 'V204'}-{matched.get('guia') or matched.get('id')}"
        pin = str(matched.get('pickup_code') or '0909')
        ose_id = str(matched.get('id'))
        
        # Actualizar en Supabase
        update_url = f"{SUPABASE_URL}/rest/v1/pedidos?id=eq.{p['id']}"
        update_data = json.dumps({
            "shalom_numero_guia": full_guia,
            "shalom_clave_recojo": pin,
            "shalom_ose_id": ose_id,
            "registrado_shalom": True
        }).encode('utf-8')
        
        up_req = urllib.request.Request(update_url, data=update_data, headers={
            "apikey": ANON_KEY,
            "Authorization": f"Bearer {ANON_KEY}",
            "Content-Type": "application/json",
            "Prefer": "return=minimal"
        }, method="PATCH")
        
        try:
            with urllib.request.urlopen(up_req, context=ctx, timeout=10) as up_resp:
                print(f"[OK] {p['codigo_seguimiento']} (DNI: {dni}) -> Guia: {full_guia} | PIN: {pin} (OSE: {ose_id})")
                synced_count += 1
        except Exception as e:
            print(f"[-] Error actualizando pedido {p['id']}: {e}")

print(f"\n=== FIN: {synced_count} PEDIDOS ACTUALIZADOS CON GUIAS OFICIALES EN SUPABASE ===")

File: scripts/sync_all_supabase_orders_with_shalom.mjs
import { createClient } from '@supabase/supabase-js';

const SUPABASE_URL = 'https://uwmdjsxwetjvsxsdngko.supabase.co';
const SUPABASE_KEY = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InV3bWRqc3h3ZXRqdnN4c2RuZ2tvIiwicm9sZSI6ImFub24iLCJpYXQiOjE3ODY2NDE5MTEsImV4cCI6MjEwMjIxNzkxMX0.KaqryIyoe4IDQGTJD_cswZkW-wfgnMcyV9tJoWxHMq8';
const SHALOM_API_KEY = 'sk_qm4rm5ivepety4ausqnubkfegp4yr2lnqu3p4q55oc3v4yzw3oma';
const SHALOM_EMAIL = 'milagrosjanetamis@gmail.com';
const SHALOM_PASSWORD = '986398Mi$';

const supabase = createClient(SUPABASE_URL, SUPABASE_KEY);

async function syncAll() {
  console.log('[*] Descargando todas las órdenes activas de Shalom Pro...');
  const allShalomOrders = [];

  for (let page = 1; page <= 3; page++) {
    const res = await fetch(`https://api.shalom-api-peru.com/v1/orders?per_page=100&page=${page}`, {
      headers: {
        'X-API-Key': SHALOM_API_KEY,
        'X-Shalom-Email': SHALOM_EMAIL,
        'X-Shalom-Password': SHALOM_PASSWORD,
      }
    });
    const json = await res.json();
    const list = json.orders || json.data || [];
    for (const item of list) {
      if (item && item.id) allShalomOrders.push(item);
    }
  }

  // Ordenar por ID descendente (las más recientes primero)
  allShalomOrders.sort((a, b) => Number(b.id || 0) - Number(a.id || 0));
  console.log(`[*] Total órdenes en Shalom Pro: ${allShalomOrders.length}`);

  // Agrupar órdenes por DNI del destinatario
  const shalomByDni = {};
  for (const o of allShalomOrders) {
    const dni = String(o.receiver?.document || '').replace(/\D/g, '').trim();
    if (dni && dni.length >= 6) {
      if (!shalomByDni[dni]) {
        shalomByDni[dni] = [];
      }
      shalomByDni[dni].push(o);
    }
  }

  // Descargar pedidos de Supabase
  const { data: pedidos, error } = await supabase
    .from('pedidos')
    .select('id, codigo_seguimiento, shalom_ose_id, shalom_numero_guia, shalom_clave_recojo, registrado_shalom, destino_detalle');

  if (error) {
    console.error('[-] Error obteniendo pedidos:', error);
    return;
  }

  console.log(`[*] Procesando ${pedidos.length} pedidos de Supabase...`);

  let updatedCount = 0;

  for (const pedido of pedidos) {
    // Extraer DNI del destino_detalle o del pedido
    const dniMatch = (pedido.destino_detalle || '').match(/DNI[^:\d]*:\s*([0-9]{8,11})/i) ||
                     (pedido.destino_detalle || '').match(/\b([0-9]{8})\b/);
    const dni = dniMatch ? dniMatch[1] : '';

    if (!dni) continue;

    const shalomOrders = shalomByDni[dni];
    if (shalomOrders && shalomOrders.length > 0) {
      // Tomar la orden MÁS RECIENTE activa (ID más alto)
      const newestShalomOrder = shalomOrders[0];
      const fullGuia = `${newestShalomOrder.serie || ''}-${newestShalomOrder.guia || ''}`;
      const oseId = String(newestShalomOrder.id);
      const pin = newestShalomOrder.pickup_code || newestShalomOrder.codigo || null;

      // Verificar si necesita actualización
      const needsUpdate = 
        pedido.shalom_ose_id !== oseId ||
        pedido.shalom_numero_guia !== fullGuia ||
        !pedido.registrado_shalom;

      if (needsUpdate) {
        console.log(`[+] Actualizando Pedido #${pedido.codigo_seguimiento} (DNI ${dni}):`);
        console.log(`    Antiguo: OSE=${pedido.shalom_ose_id} | Guía=${pedido.shalom_numero_guia}`);
        console.log(`    NUEVO (Más Reciente): OSE=${oseId} | Guía=${fullGuia} | PIN=${pin}`);

        const { error: updErr } = await supabase
          .from('pedidos')
          .update({
            shalom_ose_id: oseId,
            shalom_numero_guia: fullGuia,
            ...(pin ? { shalom_clave_recojo: pin } : {}),
            registrado_shalom: true,
          })
          .eq('id', pedido.id);

        if (updErr) {
          console.error(`    [-] Error actualizando pedido ${pedido.id}:`, updErr);
        } else {
          updatedCount++;
        }
      }
    }
  }

  console.log(`\n[✓] Sincronización completada. ${updatedCount} pedidos actualizados con las guías más recientes.`);
}

syncAll().catch(console.error);
